### ISSUE TITLE
RecursionError when parsing large (0.3 gb) sqlite file

### ISSUE THREAD

**Issue by hugolytics** [uktrade/stream-sqlite#2]
So im parsing a sqlite file from a IOT device which was sent through a post request.

```
import io
from stream_sqlite import stream_sqlite

def tables_list(table_iter):
    return [
        (table_name, table_info, list(table_rows))
        for table_name, table_info, table_rows in table_iter
        ]

tables = tables_list(stream_sqlite(io.BytesIO(sqlite_bytes),max_buffer_size=10*1024*1024))
```
gives the following (very long, which I shortened) error:
  ```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 441, in stream_sqlite
    for (name, info), single_table_rows in grouped_by_table:
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 438, in <genexpr>
    table_rows = ((table_name, table_info, row) for table_name, table_info, row in table_rows_including_internal if not table_name.startswith('sqlite_'))
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 425, in yield_table_rows
    yield from process_page(page_bytes, page_reader)
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 325, in process_table_page
    yield from \
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 304, in process_table_leaf_non_master
    yield from process_initial_payload(
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 182, in process_initial_payload
    yield from process_if_buffered_or_remember(partial(
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 411, in process_if_buffered_or_remember
    yield from process(page_bytes, page_reader)
yield from process_if_buffered_or_remember(partial(
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 411, in process_if_buffered_or_remember
    yield from process(page_bytes, page_reader)
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 200, in process_overflow_page
    yield from process_if_buffered_or_remember(partial(
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 411, in process_if_buffered_or_remember
    yield from process(page_bytes, page_reader)
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 200, in process_overflow_page
    yield from process_if_buffered_or_remember(partial(
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 411, in process_if_buffered_or_remember
    yield from process(page_bytes, page_reader)
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 200, in process_overflow_page
    yield from process_if_buffered_or_remember(partial(
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 411, in process_if_buffered_or_remember
    yield from process(page_bytes, page_reader)
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 200, in process_overflow_page
    yield from process_if_buffered_or_remember(partial(
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 411, in process_if_buffered_or_remember
    yield from process(page_bytes, page_reader)
  File ".../.venv/lib/python3.8/site-packages/stream_sqlite.py", line 189, in process_overflow_page
    num_this_page = min(payload_remainder, len(page_bytes) - 4)
RecursionError: maximum recursion depth exceeded in comparison
```

**Comment by michalc** [comment]
A few questions:

- What's the schema of this database?
- Does it contain long blobs or text strings? And if so, how long?
- Could you make this database file, or one like it that has the same error, available publicly?

**Comment by hugolytics** [comment]
I am going to have to give the worst answers to all of these questions I'm afraid...

* There is no defined schema for this database, the database is created by a their custom C++ program (which basically functions as their ORM). Maybe there is a schema, but there is no datamodel (which I wanted to define using SQLalchemy, but I can't use a sqlite connection with the database in a stream).
* Yes it contains long blobs and strings, but I don't know exactly how long yet.
* unfortunately I can't make the file publically available right now, but I can ask next week (I'm an outside consultant).

Would saving the bytes to a namedtemporarydirectory be a better option? If i can connect a sqlite cursor i could use sqlalchemy.

I also tried using [Kaitai](https://github.com/Vector35/kaitai/blob/master/sqlite3.py), but there is very little documentation, so i dont now whether that would allow me to use a validated datamodel.

On the topic of parsing with stream-slite, i tried the following:
```
data = stream_sqlite(io.BytesIO(oceon_bytes),max_buffer_size=10*1024*1024)
metadata_table = next(data)
temperature_table = next(data)
protocol_table = next(data)
chip_table = next(data)
force_scalar_table = next(data)
fluorescence_selections_metadata_table = next(data)
xy_stage_table = next(data)
camera_table = next(data)
z_stage_table = next(data)
signal_sweeps_table = next(data)
fluorescence_selections_table = next(data)
monolayer_images_table = next(data)
```
The following `next(data)` gives the same `RecursionError`.
That could be the offending table in the database.

Any thoughts?

**Comment by michalc** [comment]
> Would saving the bytes to a namedtemporarydirectory be a better option? If i can connect a sqlite cursor i could use sqlalchemy.

If you are happy to have all the bytes of the database in memory/on disk, then you don't really need this library. You have options:

- Yes saving to a temporary file, and then using sqlalchemy, or even Python's built-in sqlite3 module.
- Or accessing it directly from memory. Slightly annoyingly, I don't think Python's built-in module exposes this functionality. But APSW does https://rogerbinns.github.io/apsw/connection.html?highlight=serialize#apsw.Connection.deserialize (I tend to install APSW by the not-officially-related package apsw-wheels)

> Yes it contains long blobs and strings, but I don't know exactly how long yet.

Also on the recursion error, it might be a design limitation of this library -  so a bug, but unfortunately one that won't be quickly fixed. Long blobs and strings are split into multiple pages in the data (by default 4kb) and processing each page is done recursively. If you have a long blob, it might hit a recursion error like you've seen. I think Python's default recursion limit is 1000, so even blobs ~4mb might hit this.

(If a version could be made public, that would be useful to see if indeed it is this, or whether it is a much more fixable bug)

**Comment by hugolytics** [comment]
wow, thanks very much for your suggestions!
ill probably save to tempfile as a quick fix, and get back to stream processing when i'll optimise performance/memory use.
when i do (probably in a month or two)), ill get back to this issue. I could also increase the recursion limit, but that doesnt seem like the best thing to do.
anyway, thanks again for your help and suggestions!

**Comment by michalc** [comment]
I've made a very small start on this - have reproduced the error in a failing test in https://github.com/uktrade/stream-sqlite/pull/4

**Comment by michalc** [comment]
This I think is now fixed and released in v0.0.39
